stdlib/public/core/Duration.swift

```diff
@@ -0,0 +1,206 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@frozen
+public struct Duration: Sendable {
+  @usableFromInline
+  internal var _attoseconds: _DoubleWidth<Int64>
+  
+  @inlinable
+  internal init(_attoseconds: _DoubleWidth<Int64>) {
+    self._attoseconds = _attoseconds
+  }
+  
+  internal init(_seconds: Int64, nanoseconds: Int64) {
+    self = Duration.seconds(_seconds) + Duration.nanoseconds(nanoseconds)
+  }
+  
+  @inlinable
+  public var seconds: Int64 {
+    Int64(_attoseconds / 1_000_000_000_000_000_000)
+  }
+  
+  @inlinable
+  public var nanoseconds: Int64 {
+    let seconds = _attoseconds / 1_000_000_000_000_000_000
+    let nanoseconds = 
+      Int64((_attoseconds - seconds * 1_000_000_000_000_000_000) / 1_000_000_000)
+    return nanoseconds
+  }
+}
+
+extension Duration {
+  @inlinable
+  public static func seconds<T: BinaryInteger>(_ seconds: T) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(seconds) * 
+                                 1_000_000_000_000_000_000)
+  }
+  
+  @inlinable
+  public static func seconds(_ seconds: Double) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(seconds * 
+                                 1_000_000_000_000_000_000))
+  }
+  
+  @inlinable
+  public static func milliseconds<T: BinaryInteger>(
+    _ milliseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(milliseconds) * 
+                                 1_000_000_000_000_000)
+  }
+  
+  @inlinable
+  public static func milliseconds(_ milliseconds: Double) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(milliseconds * 
+                                 1_000_000_000_000_000))
+  }
+  
+  @inlinable
+  public static func microseconds<T: BinaryInteger>(
+    _ microseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(microseconds) * 
+                                 1_000_000_000_000)
+  }
+  
+  @inlinable
+  public static func microseconds(_ microseconds: Double) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(microseconds * 
+                                 1_000_000_000_000))
+  }
+  
+  @inlinable
+  public static func nanoseconds<T: BinaryInteger>(
+    _ nanoseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(nanoseconds) * 
+                                 1_000_000_000)
+  }
+}
+
+extension Duration: Codable {
+  private enum CodingKeys: String, CodingKey {
+    case seconds = "seconds"
+    case nanoseconds = "nanoseconds"
+  }
+  
+  public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    let seconds = try container.decode(Int64.self, forKey: .seconds)
+    let nanoseconds = try container.decode(Int64.self, forKey: .nanoseconds)
```

Comment:
The encoding here needs to be updated for the improved resolution -- we shouldn't lose information when serializing+deserializing a `Duration`.